pkg/testworkflows/testworkflowprocessor/action/containerize.go

```diff
@@ -0,0 +1,147 @@
+	constants2 "github.com/kubeshop/testkube/cmd/testworkflow-init/constants"
+	"github.com/kubeshop/testkube/pkg/testworkflows/testworkflowprocessor/action/actiontypes"
+	"github.com/kubeshop/testkube/pkg/testworkflows/testworkflowprocessor/constants"
+	stage2 "github.com/kubeshop/testkube/pkg/testworkflows/testworkflowprocessor/stage"
+)
+
+func CreateContainer(groupId int, defaultContainer stage2.Container, actions []actiontypes.Action) (cr corev1.Container, actionsCleanup []actiontypes.Action, err error) {
+	actions = slices.Clone(actions)
+	actionsCleanup = actions
+
+	// Find the container configurations and executable/setup steps
+	var setup *actiontypes.Action
+	executable := map[string]bool{}
+	containerConfigs := make([]*actiontypes.Action, 0)
+	for i := range actions {
+		if actions[i].Container != nil {
+			containerConfigs = append(containerConfigs, &actions[i])
+		} else if actions[i].Setup != nil {
+			setup = &actions[i]
+		} else if actions[i].Execute != nil {
+			executable[actions[i].Execute.Ref] = true
+		}
+	}
+
+	// Find the highest priority container configuration
+	var bestContainerConfig *actiontypes.Action
+	for i := range containerConfigs {
+		if executable[containerConfigs[i].Container.Ref] {
+			bestContainerConfig = containerConfigs[i]
+			break
+		}
+	}
+	if bestContainerConfig == nil && len(containerConfigs) > 0 {
+		bestContainerConfig = containerConfigs[len(containerConfigs)-1]
+	}
+	if bestContainerConfig == nil {
+		bestContainerConfig = &actiontypes.Action{Container: &actiontypes.ActionContainer{Config: defaultContainer.ToContainerConfig()}}
+	}
+
+	// Build the cr base
+	// TODO: Handle the case when there are multiple exclusive execution configurations
+	// TODO: Handle a case when that configuration should join multiple configurations (i.e. envs/volumeMounts)
+	if len(containerConfigs) > 0 {
+		cr, err = stage2.NewContainer().ApplyCR(&bestContainerConfig.Container.Config).ToKubernetesTemplate()
+		if err != nil {
+			return corev1.Container{}, nil, err
+		}
+
+		// Combine environment variables from each execution
+		cr.Env = nil
+		cr.EnvFrom = nil
+		for i := range containerConfigs {
+			// TODO: Avoid having multiple copies of the same environment variable
+			for _, e := range containerConfigs[i].Container.Config.Env {
+				newEnv := *e.DeepCopy()
+				computed := strings.Contains(newEnv.Value, "{{")
+				sensitive := newEnv.ValueFrom != nil && newEnv.ValueFrom.SecretKeyRef != nil
+				newEnv.Name = actiontypes.EnvName(fmt.Sprintf("%d", i), computed, sensitive, e.Name)
+				cr.Env = append(cr.Env, newEnv)
+			}
+			for _, e := range containerConfigs[i].Container.Config.EnvFrom {
+				newEnvFrom := *e.DeepCopy()
+				sensitive := newEnvFrom.SecretRef != nil
+				newEnvFrom.Prefix = actiontypes.EnvName(fmt.Sprintf("%d", i), false, sensitive, e.Prefix)
+				cr.EnvFrom = append(cr.EnvFrom, newEnvFrom)
+			}
+		}
+		// TODO: Combine the rest
```

Comment:
is it important?

Reply:
It's for later - just a comment to not forget while implementing later the merge of multiple executions in a single container 🙂